ipc-queue/src/interface_sync.rs

```diff
@@ -0,0 +1,347 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use super::*;
+use crate::fifo::*;
+use fortanix_sgx_abi::FifoDescriptor;
+
+unsafe impl<T: Send, S: Send> Send for Sender<T, S> {}
+unsafe impl<T: Send, S: Sync> Sync for Sender<T, S> {}
+
+impl<T, S: Clone> Clone for Sender<T, S> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            synchronizer: self.synchronizer.clone(),
+        }
+    }
+}
+
+impl<T: Identified, S: Synchronizer> Sender<T, S> {
+    /// Create a `Sender` from a `FifoDescriptor` and `Synchronizer`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure the following:
+    ///
+    /// * The `data` and `len` fields in `FifoDescriptor` must adhere to all
+    ///   safety requirements described in `std::slice::from_raw_parts_mut()`
+    ///
+    /// * The `offsets` field in `FifoDescriptor` must be non-null and point
+    ///   to a valid memory location holding an `AtomicUsize`.
+    ///
+    /// * The synchronizer must somehow know how to correctly synchronize with
+    ///   the other end of the channel.
+    pub unsafe fn from_descriptor(d: FifoDescriptor<T>, synchronizer: S) -> Self {
```

Comment:
Continuing the discussion from https://github.com/fortanix/rust-sgx/pull/246#discussion_r490111910 I still don't see the SGX safe user memory types being used here.

Reply:
I initially assumed you only meant the traits (`UserSafe`, `UserSafeSized`), but yeah we could also use `User<[T]>` here to check the pointer and len.